{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4Ea9zYNE6Cou"
      },
      "source": [
        "# XX\n",
        " - **Project:** GP2 Parkinson's disease meta-GWAS on European, Ashkenazi Jewish, Finnish, and Icelandic individuals \n",
        " - **Versions:** R\n",
        " - **Last Updated:** January 2025\n",
        "\n",
        "### Notebook Overview\n",
        "xx\n",
        "\n",
        "### CHANGELOG\n",
        "* 06-MAR-2025: Notebook cleaned \n",
        "* 10-FEB-2025: Notebook started"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "## Import the necessary packages \n",
        "import os\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import math\n",
        "import numbers\n",
        "import sys\n",
        "import subprocess\n",
        "import statsmodels.api as sm\n",
        "import scipy\n",
        "from scipy import stats\n",
        "from scipy.stats import norm\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "from IPython.display import display\n",
        "from tqdm import tqdm\n",
        "\n",
        "## Print out package versions\n",
        "## Getting packages loaded into this notebook and their versions to allow for reproducibility\n",
        "    # Repurposed code from stackoverflow here: https://stackoverflow.com/questions/40428931/package-for-listing-version-of-packages-used-in-a-jupyter-notebook\n",
        "\n",
        "## Import packages \n",
        "import pkg_resources\n",
        "import types\n",
        "from datetime import date\n",
        "today = date.today()\n",
        "date = today.strftime(\"%d-%b-%Y\").upper()\n",
        "\n",
        "## Define function \n",
        "def get_imports():\n",
        "    for name, val in globals().items():\n",
        "        if isinstance(val, types.ModuleType):\n",
        "            # Split ensures you get root package, not just imported function\n",
        "            name = val.__name__.split(\".\")[0]\n",
        "\n",
        "        elif isinstance(val, type):\n",
        "            name = val.__module__.split(\".\")[0]\n",
        "\n",
        "        # Some packages are weird and have different imported names vs. system/pip names\n",
        "        # Unfortunately, there is no systematic way to get pip names from a package's imported name. You'll have to add exceptions to this list manually!\n",
        "        poorly_named_packages = {\n",
        "            \"PIL\": \"Pillow\",\n",
        "            \"sklearn\": \"scikit-learn\"\n",
        "        }\n",
        "        if name in poorly_named_packages.keys():\n",
        "            name = poorly_named_packages[name]\n",
        "\n",
        "        yield name\n",
        "\n",
        "## Get a list of packages imported \n",
        "imports = list(set(get_imports()))\n",
        "\n",
        "# The only way I found to get the version of the root package from only the name of the package is to cross-check the names of installed packages vs. imported packages\n",
        "requirements = []\n",
        "for m in pkg_resources.working_set:\n",
        "    if m.project_name in imports and m.project_name!=\"pip\":\n",
        "        requirements.append((m.project_name, m.version))\n",
        "\n",
        "## Print out packages and versions \n",
        "print(f\"PACKAGE VERSIONS ({date})\")\n",
        "for r in requirements:\n",
        "    print(\"\\t{}=={}\".format(*r))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Calculate ORs and CIs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "def calculate_ors_cis(input_file_path, work_dir, output_dir):\n",
        "    with open(input_file_path, 'r') as file:\n",
        "        file_paths = [line.strip().format(WORK_DIR=work_dir) for line in file.readlines()]\n",
        "\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "    for file_path in file_paths:\n",
        "        data = pd.read_csv(file_path, sep='\\t')\n",
        "        \n",
        "        if 'BETA' in data.columns and 'SE' in data.columns:\n",
        "            # Calculate the odds ratio and the confidence intervals\n",
        "            data['OR'] = np.exp(data['BETA'])\n",
        "            data['L95'] = np.exp(data['BETA'] - 1.96 * data['SE'])\n",
        "            data['U95'] = np.exp(data['BETA'] + 1.96 * data['SE'])\n",
        "            \n",
        "            base_name = os.path.basename(file_path)\n",
        "            output_file_path = os.path.join(output_dir, base_name.replace('.txt', '.ORs.txt'))\n",
        "            \n",
        "            print(f\"Processed: {output_file_path}\")\n",
        "            data.to_csv(output_file_path, sep='\\t', index=False)\n",
        "        else:\n",
        "            print(f\"Warning: 'BETA' and 'SE' columns not found in {file_path}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "## Run through files \n",
        "calculate_ors_cis(f'{FP_DIR}/data/paths_21cohorts.txt', f'{WORK_DIR}', f'{FP_DIR}/data/processed/')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Get the summary from the final file "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "def calculate_ors_cis_meta(input_file_path, output_file_path):\n",
        "    # Load the data from the input file\n",
        "    data = pd.read_csv(input_file_path, sep='\\t')\n",
        "    \n",
        "    # Check if the required columns are present\n",
        "    if 'Effect' in data.columns and 'StdErr' in data.columns:\n",
        "        # Calculate the odds ratio and the confidence intervals\n",
        "        data['OR'] = np.exp(data['Effect'])\n",
        "        data['L95'] = np.exp(data['Effect'] - 1.96 * data['StdErr'])\n",
        "        data['U95'] = np.exp(data['Effect'] + 1.96 * data['StdErr'])\n",
        "        \n",
        "        # Save the updated dataframe to the output file\n",
        "        os.makedirs(os.path.dirname(output_file_path), exist_ok=True)\n",
        "        data.to_csv(output_file_path, sep='\\t', index=False)\n",
        "        \n",
        "        print(f\"Processed: {output_file_path}\")\n",
        "    else:\n",
        "        print(f\"Warning: 'Effect' and 'StdErr' columns not found in {input_file_path}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "## Run through files \n",
        "calculate_ors_cis_meta(f'{WORK_DIR}/METAS/METAANALYSIS1.TBL', f'{FP_DIR}/data/processed/METAANALYSIS1.TBL.ORs.txt')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Extract rsID "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SNP\tCHR\tBP\tA1\tA2\tBETA\tSE\tP\tNMISS\tOR\tL95\tU95\n",
            "chr1:596477:G:A\t1\t596477\tA\tG\t-0.0080321701243519\t0.094380997121334\t0.93211\t18429.84873202498\t0.9920000015603291\t0.8244663713923294\t1.19357688468705\n",
            "chr1:598745:C:CAG\t1\t598745\tCAG\tC\t0.0079681696370244\t0.0569300018250942\t0.8827\t18429.84873202498\t1.0079999999877503\t0.9015726659908295\t1.1269906889409174\n",
            "chr1:633423:C:T\t1\t633423\tT\tC\t0.1638180017471313\t0.0564690008759498\t0.0038085\t18429.84873202498\t1.1779999016578975\t1.054575967960476\t1.315868946823966\n",
            "chr1:701203:G:T\t1\t701203\tT\tG\t-0.2344570010900497\t0.1253699958324432\t0.061086\t18429.84873202498\t0.7910002453084649\t0.6186706641215313\t1.011331915933779\n",
            "chr1:702040:C:T\t1\t702040\tT\tC\t0.027615200728178\t0.0546800009906291\t0.61888\t18429.84873202498\t1.0280000346386708\t0.923524822286076\t1.1442941713262944\n",
            "chr1:710225:T:A\t1\t710225\tA\tT\t-0.0020020001102238\t0.042833000421524\t0.9603\t18429.84873202498\t0.9980000025553284\t0.9176358166737645\t1.08540227724626\n",
            "chr1:727233:G:A\t1\t727233\tA\tG\t0.0198026001453399\t0.0540100000798702\t0.71254\t18429.84873202498\t1.0199999723061437\t0.9175419298292933\t1.1338990728174114\n",
            "chr1:727242:G:A\t1\t727242\tA\tG\t-0.0555127002298831\t0.0359130017459392\t0.12158\t18429.84873202498\t0.9460000091765555\t0.8817010873459378\t1.0149879933298986\n",
            "chr1:727717:G:C\t1\t727717\tC\tG\t0.0534007996320724\t0.0367890000343322\t0.14588\t18429.84873202498\t1.054852344836453\t0.981468224285415\t1.1337233767472246\n"
          ]
        }
      ],
      "source": [
        "! head {FP_DIR}/data/processed/deCODE_CC_FOR_PLINK_MAF.ORs.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "                  SNP CHR         BP A1 A2    BETA      SE       P  NMISS  \\\n",
            "0  chr11:12541586:A:G  11   12541586  A  G -0.0016  0.0074  0.8270      0   \n",
            "1   chr4:70553471:A:G   4   70553471  A  G  0.0095  0.0083  0.2555      0   \n",
            "2  chr13:55101557:T:C  13   55101557  T  C  0.0116  0.0384  0.7621      0   \n",
            "3  chr2:223892881:C:T   2  223892881  T  C -0.1172  0.2445  0.6319      0   \n",
            "4   chr3:97729697:A:G   3   97729697  A  G  0.0646  0.0397  0.1040      0   \n",
            "\n",
            "         OR       L95       U95  \n",
            "0  0.998401  0.984025  1.012988  \n",
            "1  1.009545  0.993255  1.026103  \n",
            "2  1.011668  0.938320  1.090748  \n",
            "3  0.889407  0.550780  1.436228  \n",
            "4  1.066732  0.986875  1.153052  \n"
          ]
        }
      ],
      "source": [
        "df = pd.read_csv(f'{FP_DIR}/data/processed/METAANALYSIS1.TBL.ORs.txt', sep=\"\\t\")\n",
        "\n",
        "df = df.rename(columns={\n",
        "    'MarkerName': 'SNP',\n",
        "    'Allele1': 'A1',\n",
        "    'Allele2': 'A2',\n",
        "    'Effect': 'BETA',\n",
        "    'StdErr': 'SE',\n",
        "    'P-value': 'P',\n",
        "    'OR': 'OR',\n",
        "    'L95': 'L95',\n",
        "    'U95': 'U95'\n",
        "})\n",
        "\n",
        "df['CHR'] = df['SNP'].apply(lambda x: x.split(':')[0][3:])\n",
        "df['BP'] = df['SNP'].apply(lambda x: int(x.split(':')[1]))\n",
        "\n",
        "df['A1'] = df['A1'].str.upper()\n",
        "df['A2'] = df['A2'].str.upper()\n",
        "df['NMISS'] = 0 ## placeholder to keep structure \n",
        "\n",
        "df = df[['SNP', 'CHR', 'BP', 'A1', 'A2', 'BETA', 'SE', 'P', 'NMISS', 'OR', 'L95', 'U95']]\n",
        "\n",
        "# Print the dataframe to check\n",
        "print(df.head())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "df.to_csv(f'{FP_DIR}/data/processed/METAANALYSIS.processed.ORs.txt', index=False, sep=\"\\t\")"
      ]
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "RSIDS_FILE=\"${FP_DIR}/data/rsIDs_hg38_variantIDs_cojo.txt\"\n",
        "INPUT_DIR=\"${FP_DIR}/data/processed/\"\n",
        "OUTPUT_FILE=\"${FP_DIR}/data/rsIDs_across_cohorts.txt\"\n",
        "\n",
        "# Clear the output file if it already exists\n",
        "> $OUTPUT_FILE\n",
        "\n",
        "# Flag to track whether the header has been written\n",
        "HEADER_WRITTEN=false\n",
        "\n",
        "# Loop through each SNP in the rsIDs file and use grep to find matches in the input files\n",
        "while IFS= read -r SNP\n",
        "do\n",
        "    # Search for matching rows in all .ORs.txt files\n",
        "    for file in ${INPUT_DIR}/*.ORs.txt; do\n",
        "        # Extract filename without the extension\n",
        "        filename=$(basename \"$file\" .ORs.txt)\n",
        "\n",
        "        if [ \"$HEADER_WRITTEN\" = false ]; then\n",
        "            # Write header from the first file only if it hasn't been written yet, and add a new column header for \"Filename\"\n",
        "            head -n 1 \"$file\" | awk '{print $0 \"\\tFilename\"}' >> $OUTPUT_FILE\n",
        "            HEADER_WRITTEN=true\n",
        "        fi\n",
        "        \n",
        "        # Extract matching rows by SNP, excluding the header, and add the filename to each row\n",
        "        grep -h -w \"$SNP\" \"$file\" | grep -v \"^rsID\" | awk -v fname=\"$filename\" '{print $0 \"\\t\" fname}' >> $OUTPUT_FILE\n",
        "    done\n",
        "done < <(cut -f2 $RSIDS_FILE)  "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "import os\n",
        "import pandas as pd\n",
        "from tqdm import tqdm\n",
        "\n",
        "# Define the file paths\n",
        "rsids_file = f\"{FP_DIR}/data/rsIDs_hg38_variantIDs_cojo.txt\"\n",
        "input_file = f\"{FP_DIR}/data/rsIDs_across_cohorts.txt\"\n",
        "cohort_names_file = f\"{FP_DIR}/data/cohort_names.txt\"\n",
        "\n",
        "output_file = f\"{FP_DIR}/data/rsIDs_cojo_with_ORs_allCohorts.txt\"\n",
        "split_output_dir = f\"{FP_DIR}/data/rsID_files_withMeta_cojo/\"\n",
        "\n",
        "# Create output directory for split files if it doesn't exist\n",
        "os.makedirs(split_output_dir, exist_ok=True)\n",
        "\n",
        "# Function to clean and standardize filenames\n",
        "def clean_filename(filename):\n",
        "    return filename.strip().replace(\" \", \"_\").replace(\"'\", \"\")\n",
        "\n",
        "# Read in data files\n",
        "rsids_df = pd.read_csv(rsids_file, sep=\"\\t\")\n",
        "input_df = pd.read_csv(input_file, sep=\"\\t\")\n",
        "cohort_map_df = pd.read_csv(cohort_names_file, sep=\"\\t\")\n",
        "\n",
        "# Create a mapping dictionary for Filename -> Rename\n",
        "filename_to_cohort_map = dict(zip(cohort_map_df['Filename'], cohort_map_df['Rename']))\n",
        "\n",
        "# Filter the input DataFrame to keep only rows with SNPs that are in the snp_to_rsid_map\n",
        "filtered_df = input_df[input_df['SNP'].isin(dict(zip(rsids_df['SNP'], rsids_df['rsID'])).keys())].copy()\n",
        "filtered_df['rsID'] = filtered_df['SNP'].map(dict(zip(rsids_df['SNP'], rsids_df['rsID'])))\n",
        "\n",
        "# Map the old filenames to new cohort names\n",
        "filtered_df['Cohort'] = filtered_df['Filename'].map(filename_to_cohort_map)\n",
        "\n",
        "# Select the desired columns and reorder them\n",
        "desired_columns = ['rsID', 'SNP', 'Cohort', 'BETA', 'SE', 'P', 'OR', 'L95', 'U95']\n",
        "reordered_df = filtered_df[desired_columns]\n",
        "reordered_df.to_csv(output_file, sep=\"\\t\", index=False)\n",
        "\n",
        "# Loop through each unique rsID and save separate files for each\n",
        "for rsid, group_df in tqdm(reordered_df.groupby('rsID'), desc=\"Saving individual rsID files\", total=reordered_df['rsID'].nunique()):\n",
        "    cleaned_rsid = clean_filename(rsid)  # Clean the rsID for consistent filenames\n",
        "    output_split_file = os.path.join(split_output_dir, f\"{cleaned_rsid}_split.txt\")\n",
        "    group_df.to_csv(output_split_file, sep=\"\\t\", index=False)\n",
        "\n",
        "# Final messages\n",
        "print(f\"Processing complete. Filtered and reordered file saved as: {output_file}\")\n",
        "print(f\"Individual rsID files saved in: {split_output_dir}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "155\n"
          ]
        }
      ],
      "source": [
        "! ls {FP_DIR}/data/rsID_files_withMeta_cojo | wc -l"
      ]
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "## Going back for the HLA variant\n",
        "VARIANT=\"chr6:32603345:C:T\"\n",
        "RSID=\"rs2647066\"\n",
        "INPUT_DIR=\"${FP_DIR}/data/processed/\"\n",
        "OUTPUT_FILE=\"${FP_DIR}/data/variant_${RSID}_across_cohorts.txt\"\n",
        "\n",
        "# Clear the output file if it already exists\n",
        "> $OUTPUT_FILE\n",
        "\n",
        "# Flag to track whether the header has been written\n",
        "HEADER_WRITTEN=false\n",
        "\n",
        "# Loop through each input file to find matches for the specified variant\n",
        "for file in ${INPUT_DIR}/*.ORs.txt; do\n",
        "    # Extract filename without the extension\n",
        "    filename=$(basename \"$file\" .ORs.txt)\n",
        "\n",
        "    if [ \"$HEADER_WRITTEN\" = false ]; then\n",
        "        # Write header from the first file only if it hasn't been written yet, and add a new column header for \"Filename\"\n",
        "        head -n 1 \"$file\" | awk '{print $0 \"\\tFilename\"}' >> $OUTPUT_FILE\n",
        "        HEADER_WRITTEN=true\n",
        "    fi\n",
        "    \n",
        "    # Extract matching rows by variant, excluding the header, and add the filename to each row\n",
        "    grep -h -w \"$VARIANT\" \"$file\" | grep -v \"^rsID\" | awk -v fname=\"$filename\" '{print $0 \"\\t\" fname}' >> $OUTPUT_FILE\n",
        "done"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "%%bash\n",
        "\n",
        "echo \"./GP2/POP_CONTROLS/GP2_EUR_CC_FOR_PLINK_MAF_AGE_POP_CONTROLS.txt\"\n",
        "head -1 ./GP2/POP_CONTROLS/GP2_EUR_CC_FOR_PLINK_MAF_AGE_POP_CONTROLS.txt\n",
        "grep \"chr12:40340400:G:A\" ./GP2/POP_CONTROLS/GP2_EUR_CC_FOR_PLINK_MAF_AGE_POP_CONTROLS.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_COURAGE_UK_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_COURAGE_UK_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_COURAGE_UK_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_DUTCH_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_DUTCH_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_DUTCH_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_GERMANY_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_GERMANY_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_GERMANY_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_MCGILL_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_MCGILL_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_MCGILL_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_MF_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_MF_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_MF_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_NIA_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_NIA_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_NIA_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_OSLO_EUR_CC_FOR_PLINK_PD.txt\"\n",
        "head -1 ./IPDGC/IPDGC_OSLO_EUR_CC_FOR_PLINK_PD.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_OSLO_EUR_CC_FOR_PLINK_PD.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_SHULMAN_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_SHULMAN_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_SHULMAN_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_SPAIN3_EUR_CC_NO_AGE_COVAR_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_SPAIN3_EUR_CC_NO_AGE_COVAR_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_SPAIN3_EUR_CC_NO_AGE_COVAR_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_TUBI_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_TUBI_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_TUBI_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_UK_GWAS_EUR_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./IPDGC/IPDGC_UK_GWAS_EUR_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_UK_GWAS_EUR_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./METAS/FIGS_META_EUR_MAF_FOR_PLINK.txt\"\n",
        "head -1 ./METAS/FIGS_META_EUR_MAF_FOR_PLINK.txt\n",
        "grep \"chr12:40340400:G:A\" ./METAS/FIGS_META_EUR_MAF_FOR_PLINK.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./UKB/UKB_ALL/UKB_ALL_CC_FOR_PLINK_MAF\"\n",
        "head -1 ./UKB/UKB_ALL/UKB_ALL_CC_FOR_PLINK_MAF\n",
        "grep \"chr12:40340400:G:A\" ./UKB/UKB_ALL/UKB_ALL_CC_FOR_PLINK_MAF\n",
        "echo \"\"\n",
        "\n",
        "echo \"./GP2/UPDATED/GP2_AJ_CC_FOR_PLINK_MAF_update.txt\"\n",
        "head -1 ./GP2/UPDATED/GP2_AJ_CC_FOR_PLINK_MAF_update.txt\n",
        "grep \"chr12:40340400:G:A\" ./GP2/UPDATED/GP2_AJ_CC_FOR_PLINK_MAF_update.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./FIGS/FIGS_AJ_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./FIGS/FIGS_AJ_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./FIGS/FIGS_AJ_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./FINNGEN_R10_PD/finngen_R11_G6_PARKINSON_FOR_PLINK\"\n",
        "head -1 ./FINNGEN_R10_PD/finngen_R11_G6_PARKINSON_FOR_PLINK\n",
        "grep \"chr12:40340400:G:A\" ./FINNGEN_R10_PD/finngen_R11_G6_PARKINSON_FOR_PLINK\n",
        "echo \"\"\n",
        "\n",
        "echo \"./IPDGC/IPDGC_FINLAND_FIN_CC_NO_AGE_FOR_PLINK.txt\"\n",
        "head -1 ./IPDGC/IPDGC_FINLAND_FIN_CC_NO_AGE_FOR_PLINK.txt\n",
        "grep \"chr12:40340400:G:A\" ./IPDGC/IPDGC_FINLAND_FIN_CC_NO_AGE_FOR_PLINK.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./deCODE/deCODE_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./deCODE/deCODE_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./deCODE/deCODE_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./MVP_CC_FOR_PLINK_MAF.txt\"\n",
        "head -1 ./MVP_CC_FOR_PLINK_MAF.txt\n",
        "grep \"chr12:40340400:G:A\" ./MVP_CC_FOR_PLINK_MAF.txt\n",
        "echo \"\"\n",
        "\n",
        "echo \"./UKB/UKB_PROXY/UKB_PROXY_ADJUSTED_FOR_PLINK.txt\"\n",
        "head -1 ./UKB/UKB_PROXY/UKB_PROXY_ADJUSTED_FOR_PLINK.txt\n",
        "grep \"chr12:40340400:G:A\" ./UKB/UKB_PROXY/UKB_PROXY_ADJUSTED_FOR_PLINK.txt\n",
        "echo \"\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "CHR\tBP\tSNP\tA1\tA2\tN\tP\tP(R)\tBETA\tBETA(R)\tQ\tI\tWEIGHTED_Z\tP(WZ)\tF0\tF1\tF2\tF3\tF4\tF5\tF6\tF7\tF8\tF9\tF10\tF11\tF12\tF13\tF14\tF15\tF16\tF17\tF18\tF19\tF20\n",
            "12\t40340400\tchr12:40340400:G:A\tA\tG\t8\t5.892e-89\t5.343e-30\t2.208\t2.3254\t0.0031\t67.39\t\t\t3.278\t\t\t\t\t\t\t\t\t2.1774\t\t\t3.0949\t1.8799\t2.0018\t2.1167\t\t\t\t1.6378\t2.983\n"
          ]
        }
      ],
      "source": [
        "! head -1 {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP.lrrk2_special_filter\n",
        "! grep \"chr12:40340400:G:A\" {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP.lrrk2_special_filter"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>rsID</th>\n",
              "      <th>SNP(hg38)</th>\n",
              "      <th>Cohort</th>\n",
              "      <th>BETA</th>\n",
              "      <th>P</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] GP2</td>\n",
              "      <td>3.2780</td>\n",
              "      <td>3.440000e-10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] IPDGC - Spain</td>\n",
              "      <td>2.1774</td>\n",
              "      <td>6.490000e-09</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] FIGS</td>\n",
              "      <td>3.0949</td>\n",
              "      <td>1.740000e-20</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] UKBiobank - Case/Control</td>\n",
              "      <td>1.8799</td>\n",
              "      <td>3.460000e-07</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[AJ] GP2</td>\n",
              "      <td>2.0018</td>\n",
              "      <td>9.180000e-09</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[AJ] FIGS</td>\n",
              "      <td>2.1167</td>\n",
              "      <td>2.340000e-25</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] Million Veteran Program</td>\n",
              "      <td>1.6378</td>\n",
              "      <td>1.540000e-11</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] UKBiobank - Proxy Cases</td>\n",
              "      <td>2.9830</td>\n",
              "      <td>2.040000e-12</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "         rsID           SNP(hg38)                          Cohort    BETA  \\\n",
              "0  rs34637584  chr12:40340400:G:A                       [EUR] GP2  3.2780   \n",
              "1  rs34637584  chr12:40340400:G:A             [EUR] IPDGC - Spain  2.1774   \n",
              "2  rs34637584  chr12:40340400:G:A                      [EUR] FIGS  3.0949   \n",
              "3  rs34637584  chr12:40340400:G:A  [EUR] UKBiobank - Case/Control  1.8799   \n",
              "4  rs34637584  chr12:40340400:G:A                        [AJ] GP2  2.0018   \n",
              "5  rs34637584  chr12:40340400:G:A                       [AJ] FIGS  2.1167   \n",
              "6  rs34637584  chr12:40340400:G:A   [EUR] Million Veteran Program  1.6378   \n",
              "7  rs34637584  chr12:40340400:G:A   [EUR] UKBiobank - Proxy Cases  2.9830   \n",
              "\n",
              "              P  \n",
              "0  3.440000e-10  \n",
              "1  6.490000e-09  \n",
              "2  1.740000e-20  \n",
              "3  3.460000e-07  \n",
              "4  9.180000e-09  \n",
              "5  2.340000e-25  \n",
              "6  1.540000e-11  \n",
              "7  2.040000e-12  "
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "g2019s = pd.read_csv(f'./../data/LRRK2-G2019S-rs34637584.tsv', sep=\"\\t\")\n",
        "g2019s"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>rsID</th>\n",
              "      <th>SNP(hg38)</th>\n",
              "      <th>Cohort</th>\n",
              "      <th>BETA</th>\n",
              "      <th>P</th>\n",
              "      <th>SE</th>\n",
              "      <th>OR</th>\n",
              "      <th>L95</th>\n",
              "      <th>U95</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] GP2</td>\n",
              "      <td>3.2780</td>\n",
              "      <td>3.440000e-10</td>\n",
              "      <td>0.522180</td>\n",
              "      <td>26.522674</td>\n",
              "      <td>9.530784</td>\n",
              "      <td>73.808438</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] IPDGC - Spain</td>\n",
              "      <td>2.1774</td>\n",
              "      <td>6.490000e-09</td>\n",
              "      <td>0.375180</td>\n",
              "      <td>8.823336</td>\n",
              "      <td>4.229346</td>\n",
              "      <td>18.407398</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] FIGS</td>\n",
              "      <td>3.0949</td>\n",
              "      <td>1.740000e-20</td>\n",
              "      <td>0.333603</td>\n",
              "      <td>22.085030</td>\n",
              "      <td>11.484957</td>\n",
              "      <td>42.468470</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] UKBiobank - Case/Control</td>\n",
              "      <td>1.8799</td>\n",
              "      <td>3.460000e-07</td>\n",
              "      <td>0.368861</td>\n",
              "      <td>6.552850</td>\n",
              "      <td>3.180161</td>\n",
              "      <td>13.502411</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[AJ] GP2</td>\n",
              "      <td>2.0018</td>\n",
              "      <td>9.180000e-09</td>\n",
              "      <td>0.348429</td>\n",
              "      <td>7.402368</td>\n",
              "      <td>3.739231</td>\n",
              "      <td>14.654099</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[AJ] FIGS</td>\n",
              "      <td>2.1167</td>\n",
              "      <td>2.340000e-25</td>\n",
              "      <td>0.203421</td>\n",
              "      <td>8.303690</td>\n",
              "      <td>5.573341</td>\n",
              "      <td>12.371622</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] Million Veteran Program</td>\n",
              "      <td>1.6378</td>\n",
              "      <td>1.540000e-11</td>\n",
              "      <td>0.242850</td>\n",
              "      <td>5.143841</td>\n",
              "      <td>3.195725</td>\n",
              "      <td>8.279529</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>rs34637584</td>\n",
              "      <td>chr12:40340400:G:A</td>\n",
              "      <td>[EUR] UKBiobank - Proxy Cases</td>\n",
              "      <td>2.9830</td>\n",
              "      <td>2.040000e-12</td>\n",
              "      <td>0.424220</td>\n",
              "      <td>19.746969</td>\n",
              "      <td>8.597986</td>\n",
              "      <td>45.352804</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "         rsID           SNP(hg38)                          Cohort    BETA  \\\n",
              "0  rs34637584  chr12:40340400:G:A                       [EUR] GP2  3.2780   \n",
              "1  rs34637584  chr12:40340400:G:A             [EUR] IPDGC - Spain  2.1774   \n",
              "2  rs34637584  chr12:40340400:G:A                      [EUR] FIGS  3.0949   \n",
              "3  rs34637584  chr12:40340400:G:A  [EUR] UKBiobank - Case/Control  1.8799   \n",
              "4  rs34637584  chr12:40340400:G:A                        [AJ] GP2  2.0018   \n",
              "5  rs34637584  chr12:40340400:G:A                       [AJ] FIGS  2.1167   \n",
              "6  rs34637584  chr12:40340400:G:A   [EUR] Million Veteran Program  1.6378   \n",
              "7  rs34637584  chr12:40340400:G:A   [EUR] UKBiobank - Proxy Cases  2.9830   \n",
              "\n",
              "              P        SE         OR        L95        U95  \n",
              "0  3.440000e-10  0.522180  26.522674   9.530784  73.808438  \n",
              "1  6.490000e-09  0.375180   8.823336   4.229346  18.407398  \n",
              "2  1.740000e-20  0.333603  22.085030  11.484957  42.468470  \n",
              "3  3.460000e-07  0.368861   6.552850   3.180161  13.502411  \n",
              "4  9.180000e-09  0.348429   7.402368   3.739231  14.654099  \n",
              "5  2.340000e-25  0.203421   8.303690   5.573341  12.371622  \n",
              "6  1.540000e-11  0.242850   5.143841   3.195725   8.279529  \n",
              "7  2.040000e-12  0.424220  19.746969   8.597986  45.352804  "
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "g2019s[\"SE\"] = np.abs(g2019s[\"BETA\"] / norm.ppf(g2019s[\"P\"] / 2))\n",
        "\n",
        "g2019s[\"OR\"] = np.exp(g2019s[\"BETA\"])\n",
        "g2019s[\"L95\"] = np.exp(g2019s[\"BETA\"] - 1.96 * g2019s[\"SE\"])\n",
        "g2019s[\"U95\"] = np.exp(g2019s[\"BETA\"] + 1.96 * g2019s[\"SE\"])\n",
        "\n",
        "display(g2019s)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "g2019s.to_csv('./../data/LRRK2-G2019S-rs34637584_with_SE_OR_L95_U95.tsv', sep=\"\\t\", index=False)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            " CHR          BP            SNP  A1  A2   N           P        P(R)    BETA BETA(R)       Q       I  WEIGHTED_Z       P(WZ)      F0      F1      F2      F3      F4      F5      F6      F7      F8      F9     F10     F11     F12     F13     F14     F15     F16     F17     F18     F19     F20\n",
            "   3   161328149 chr3:161328149:A:T   T   A  20   7.365e-13   2.312e-07  0.0572  0.0657  0.0252   42.12       6.408   1.479e-10  0.0844  0.2715  0.2099  0.1067 -0.0275  0.1448  0.0916  0.0742  0.0827  0.0244  0.0608  0.1455  0.0419  0.0483  0.0932  0.0833  0.0065  0.1100  0.0658      NA  0.0127\n",
            "  16    52939375 chr16:52939375:A:G   G   A  20   2.308e-09   1.975e-05  0.0537  0.0622  0.0221   42.96       5.676   1.382e-08  0.0944  0.0552  0.0745  0.2081 -0.0522  0.0552  0.1863  0.1134  0.1968  0.1190  0.1324  0.0412  0.0313  0.0474  0.0485  0.3130 -0.0114  0.1949  0.0141      NA  0.0433\n",
            "   3    58410136 chr3:58410136:A:G   G   A  20   4.216e-08   4.216e-08  0.0429  0.0429  0.5971    0.00       5.472   4.455e-08  0.0341  0.0721  0.0766  0.0054  0.1779  0.0767 -0.0009  0.0236 -0.0589  0.1219 -0.0323  0.0069  0.0535  0.0324  0.0541 -0.1088  0.0228  0.0828  0.0611      NA  0.0531\n",
            "   1   161000299 chr1:161000299:C:T   T   C  18   4.276e-10    1.09e-07  0.2103  0.2188  0.2410   17.79       5.715   1.102e-08  0.1887  0.2718  0.7690  0.4336      NA      NA -0.1117  0.8976 -0.9793      NA -0.0166  0.2000  0.2449  0.2392  0.5484  0.7396  0.1973 -0.1397  0.3716  0.0398  0.1452\n",
            "   3   151990484 chr3:151990484:CGAAA:C   C CGAAA  20   1.611e-08   1.611e-08 -0.0532 -0.0532  0.5094    0.00      -5.312   1.088e-07 -0.0478 -0.0118 -0.2628 -0.0657  0.0769 -0.1033 -0.0783 -0.1571 -0.3621  0.0054 -0.1697 -0.0099 -0.0607 -0.0780 -0.0089  0.0232 -0.0498 -0.1185 -0.0576      NA -0.0371\n",
            "   6    29015989 chr6:29015989:G:A   A   G  21   4.998e-09    8.77e-08 -0.0983 -0.0995  0.3563    7.89      -5.598   2.169e-08 -0.1399 -0.1887 -0.2933 -0.3220 -0.4462 -0.3666 -0.1609  0.1107 -0.8173 -0.1031 -0.2408 -0.0419 -0.0812 -0.1065 -0.1523  0.2035 -0.0989 -0.1220 -0.0747 -0.0488 -0.0799\n"
          ]
        }
      ],
      "source": [
        "! head -1 {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP_beta_fix.meta\n",
        "# rs11929238\n",
        "! grep \"chr3:161328149:A:T\" {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP_beta_fix.meta\n",
        "\n",
        "# rs1971534\n",
        "! grep \"chr16:52939375:A:G\" {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP_beta_fix.meta\n",
        "\n",
        "# rs56384862\n",
        "! grep \"chr3:58410136:A:G\" {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP_beta_fix.meta\n",
        "\n",
        "# rs72712862\n",
        "! grep \"chr1:161000299:C:T\" {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP_beta_fix.meta\n",
        "\n",
        "# rs10569457\n",
        "! grep \"chr3:151990484:CGAAA:C\" {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP_beta_fix.meta\n",
        "\n",
        "# rs41316625\n",
        "! grep \"chr6:29015989:G:A\" {WORK_DIR}/METAS/EUR_ALL_WITH_PROXY_META_MAF_MVP_beta_fix.meta"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>CHR</th>\n",
              "      <th>BP</th>\n",
              "      <th>SNP</th>\n",
              "      <th>A1</th>\n",
              "      <th>A2</th>\n",
              "      <th>N</th>\n",
              "      <th>P</th>\n",
              "      <th>P(R)</th>\n",
              "      <th>BETA</th>\n",
              "      <th>BETA(R)</th>\n",
              "      <th>...</th>\n",
              "      <th>[EUR] IPDGC - UK GWAS</th>\n",
              "      <th>[EUR] FIGS</th>\n",
              "      <th>[EUR] UKBiobank - Case/Control</th>\n",
              "      <th>[AJ] GP2</th>\n",
              "      <th>[AJ] FIGS</th>\n",
              "      <th>[FIN] FinnGen - R11</th>\n",
              "      <th>[FIN] IPDGC - Finland</th>\n",
              "      <th>[ICE] deCODE</th>\n",
              "      <th>[EUR] Million Veteran Program</th>\n",
              "      <th>[EUR] UKBiobank - Proxy Cases</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>3</td>\n",
              "      <td>161328149</td>\n",
              "      <td>chr3:161328149:A:T</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>20</td>\n",
              "      <td>7.365000e-13</td>\n",
              "      <td>2.312000e-07</td>\n",
              "      <td>0.0572</td>\n",
              "      <td>0.0657</td>\n",
              "      <td>...</td>\n",
              "      <td>0.1455</td>\n",
              "      <td>0.0419</td>\n",
              "      <td>0.0483</td>\n",
              "      <td>0.0932</td>\n",
              "      <td>0.0833</td>\n",
              "      <td>0.0065</td>\n",
              "      <td>0.1100</td>\n",
              "      <td>0.0658</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0127</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>16</td>\n",
              "      <td>52939375</td>\n",
              "      <td>chr16:52939375:A:G</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>20</td>\n",
              "      <td>2.308000e-09</td>\n",
              "      <td>1.975000e-05</td>\n",
              "      <td>0.0537</td>\n",
              "      <td>0.0622</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0412</td>\n",
              "      <td>0.0313</td>\n",
              "      <td>0.0474</td>\n",
              "      <td>0.0485</td>\n",
              "      <td>0.3130</td>\n",
              "      <td>-0.0114</td>\n",
              "      <td>0.1949</td>\n",
              "      <td>0.0141</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0433</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>3</td>\n",
              "      <td>58410136</td>\n",
              "      <td>chr3:58410136:A:G</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>20</td>\n",
              "      <td>4.216000e-08</td>\n",
              "      <td>4.216000e-08</td>\n",
              "      <td>0.0429</td>\n",
              "      <td>0.0429</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0069</td>\n",
              "      <td>0.0535</td>\n",
              "      <td>0.0324</td>\n",
              "      <td>0.0541</td>\n",
              "      <td>-0.1088</td>\n",
              "      <td>0.0228</td>\n",
              "      <td>0.0828</td>\n",
              "      <td>0.0611</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0531</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>161000299</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>18</td>\n",
              "      <td>4.276000e-10</td>\n",
              "      <td>1.090000e-07</td>\n",
              "      <td>0.2103</td>\n",
              "      <td>0.2188</td>\n",
              "      <td>...</td>\n",
              "      <td>0.2000</td>\n",
              "      <td>0.2449</td>\n",
              "      <td>0.2392</td>\n",
              "      <td>0.5484</td>\n",
              "      <td>0.7396</td>\n",
              "      <td>0.1973</td>\n",
              "      <td>-0.1397</td>\n",
              "      <td>0.3716</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.1452</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>3</td>\n",
              "      <td>151990484</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>C</td>\n",
              "      <td>CGAAA</td>\n",
              "      <td>20</td>\n",
              "      <td>1.611000e-08</td>\n",
              "      <td>1.611000e-08</td>\n",
              "      <td>-0.0532</td>\n",
              "      <td>-0.0532</td>\n",
              "      <td>...</td>\n",
              "      <td>-0.0099</td>\n",
              "      <td>-0.0607</td>\n",
              "      <td>-0.0780</td>\n",
              "      <td>-0.0089</td>\n",
              "      <td>0.0232</td>\n",
              "      <td>-0.0498</td>\n",
              "      <td>-0.1185</td>\n",
              "      <td>-0.0576</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0371</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows Ã— 35 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "   CHR         BP                     SNP A1     A2   N             P  \\\n",
              "0    3  161328149      chr3:161328149:A:T  T      A  20  7.365000e-13   \n",
              "1   16   52939375      chr16:52939375:A:G  G      A  20  2.308000e-09   \n",
              "2    3   58410136       chr3:58410136:A:G  G      A  20  4.216000e-08   \n",
              "3    1  161000299      chr1:161000299:C:T  T      C  18  4.276000e-10   \n",
              "4    3  151990484  chr3:151990484:CGAAA:C  C  CGAAA  20  1.611000e-08   \n",
              "\n",
              "           P(R)    BETA  BETA(R)  ...  [EUR] IPDGC - UK GWAS  [EUR] FIGS  \\\n",
              "0  2.312000e-07  0.0572   0.0657  ...                 0.1455      0.0419   \n",
              "1  1.975000e-05  0.0537   0.0622  ...                 0.0412      0.0313   \n",
              "2  4.216000e-08  0.0429   0.0429  ...                 0.0069      0.0535   \n",
              "3  1.090000e-07  0.2103   0.2188  ...                 0.2000      0.2449   \n",
              "4  1.611000e-08 -0.0532  -0.0532  ...                -0.0099     -0.0607   \n",
              "\n",
              "   [EUR] UKBiobank - Case/Control  [AJ] GP2  [AJ] FIGS  [FIN] FinnGen - R11  \\\n",
              "0                          0.0483    0.0932     0.0833               0.0065   \n",
              "1                          0.0474    0.0485     0.3130              -0.0114   \n",
              "2                          0.0324    0.0541    -0.1088               0.0228   \n",
              "3                          0.2392    0.5484     0.7396               0.1973   \n",
              "4                         -0.0780   -0.0089     0.0232              -0.0498   \n",
              "\n",
              "   [FIN] IPDGC - Finland  [ICE] deCODE  [EUR] Million Veteran Program  \\\n",
              "0                 0.1100        0.0658                            NaN   \n",
              "1                 0.1949        0.0141                            NaN   \n",
              "2                 0.0828        0.0611                            NaN   \n",
              "3                -0.1397        0.3716                            NaN   \n",
              "4                -0.1185       -0.0576                            NaN   \n",
              "\n",
              "   [EUR] UKBiobank - Proxy Cases  \n",
              "0                         0.0127  \n",
              "1                         0.0433  \n",
              "2                         0.0531  \n",
              "3                         0.1452  \n",
              "4                        -0.0371  \n",
              "\n",
              "[5 rows x 35 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "file_path = f'{WORK_DIR}/METAS/FINAL_SETS/FOREST_PLOTS/data/error_3_rsIDs.txt'\n",
        "\n",
        "# Reading the tab-delimited file\n",
        "snp_data = pd.read_csv(file_path, delimiter='\\t')\n",
        "\n",
        "# Mapping old column names to new ones\n",
        "column_renames = {\n",
        "    \"F0\": \"[EUR] GP2\",\n",
        "    \"F1\": \"[EUR] IPDGC - Courage UK\",\n",
        "    \"F2\": \"[EUR] IPDGC - Dutch\",\n",
        "    \"F3\": \"[EUR] IPDGC - Germany\",\n",
        "    \"F4\": \"[EUR] IPDGC - McGill\",\n",
        "    \"F5\": \"[EUR] IPDGC - MF\",\n",
        "    \"F6\": \"[EUR] IPDGC - NIA\",\n",
        "    \"F7\": \"[EUR] IPDGC - Oslo\",\n",
        "    \"F8\": \"[EUR] IPDGC - Shulman\",\n",
        "    \"F9\": \"[EUR] IPDGC - Spain\",\n",
        "    \"F10\": \"[EUR] IPDGC - TÃ¼bingen\",\n",
        "    \"F11\": \"[EUR] IPDGC - UK GWAS\",\n",
        "    \"F12\": \"[EUR] FIGS\",\n",
        "    \"F13\": \"[EUR] UKBiobank - Case/Control\",\n",
        "    \"F14\": \"[AJ] GP2\",\n",
        "    \"F15\": \"[AJ] FIGS\",\n",
        "    \"F16\": \"[FIN] FinnGen - R11\",\n",
        "    \"F17\": \"[FIN] IPDGC - Finland\",\n",
        "    \"F18\": \"[ICE] deCODE\",\n",
        "    \"F19\": \"[EUR] Million Veteran Program\",\n",
        "    \"F20\": \"[EUR] UKBiobank - Proxy Cases\"\n",
        "}\n",
        "\n",
        "snp_data.rename(columns=column_renames, inplace=True)\n",
        "display(snp_data.head())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>rsID</th>\n",
              "      <th>SNP(hg38)</th>\n",
              "      <th>Cohort</th>\n",
              "      <th>BETA</th>\n",
              "      <th>P</th>\n",
              "      <th>SE</th>\n",
              "      <th>OR</th>\n",
              "      <th>L95</th>\n",
              "      <th>U95</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>109</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[ICE] deCODE</td>\n",
              "      <td>-0.057629</td>\n",
              "      <td>0.062097</td>\n",
              "      <td>0.030890</td>\n",
              "      <td>0.944000</td>\n",
              "      <td>0.888541</td>\n",
              "      <td>1.002920</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[EUR] IPDGC - McGill</td>\n",
              "      <td>0.076878</td>\n",
              "      <td>0.482758</td>\n",
              "      <td>0.109532</td>\n",
              "      <td>1.079910</td>\n",
              "      <td>0.871268</td>\n",
              "      <td>1.338515</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[EUR] IPDGC - NIA</td>\n",
              "      <td>-0.078313</td>\n",
              "      <td>0.336272</td>\n",
              "      <td>0.081444</td>\n",
              "      <td>0.924675</td>\n",
              "      <td>0.788248</td>\n",
              "      <td>1.084715</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>91</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[AJ] FIGS</td>\n",
              "      <td>0.023228</td>\n",
              "      <td>0.780758</td>\n",
              "      <td>0.083455</td>\n",
              "      <td>1.023500</td>\n",
              "      <td>0.869060</td>\n",
              "      <td>1.205386</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[EUR] IPDGC - Germany</td>\n",
              "      <td>-0.065666</td>\n",
              "      <td>0.536008</td>\n",
              "      <td>0.106107</td>\n",
              "      <td>0.936444</td>\n",
              "      <td>0.760609</td>\n",
              "      <td>1.152928</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] UKBiobank - Case/Control</td>\n",
              "      <td>0.239174</td>\n",
              "      <td>0.016003</td>\n",
              "      <td>0.099290</td>\n",
              "      <td>1.270200</td>\n",
              "      <td>1.045573</td>\n",
              "      <td>1.543085</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>66</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] IPDGC - UK GWAS</td>\n",
              "      <td>0.200022</td>\n",
              "      <td>0.308008</td>\n",
              "      <td>0.196214</td>\n",
              "      <td>1.221430</td>\n",
              "      <td>0.831473</td>\n",
              "      <td>1.794275</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] FIGS</td>\n",
              "      <td>0.244900</td>\n",
              "      <td>0.049260</td>\n",
              "      <td>0.124546</td>\n",
              "      <td>1.277494</td>\n",
              "      <td>1.000789</td>\n",
              "      <td>1.630703</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[ICE] deCODE</td>\n",
              "      <td>0.371564</td>\n",
              "      <td>0.000516</td>\n",
              "      <td>0.107008</td>\n",
              "      <td>1.450001</td>\n",
              "      <td>1.175658</td>\n",
              "      <td>1.788361</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>60</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] IPDGC - TÃ¼bingen</td>\n",
              "      <td>-0.016584</td>\n",
              "      <td>0.975027</td>\n",
              "      <td>0.529763</td>\n",
              "      <td>0.983553</td>\n",
              "      <td>0.348221</td>\n",
              "      <td>2.778054</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>122 rows Ã— 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "           rsID               SNP(hg38)                          Cohort  \\\n",
              "109  rs10569457  chr3:151990484:CGAAA:C                    [ICE] deCODE   \n",
              "27   rs10569457  chr3:151990484:CGAAA:C            [EUR] IPDGC - McGill   \n",
              "38   rs10569457  chr3:151990484:CGAAA:C               [EUR] IPDGC - NIA   \n",
              "91   rs10569457  chr3:151990484:CGAAA:C                       [AJ] FIGS   \n",
              "22   rs10569457  chr3:151990484:CGAAA:C           [EUR] IPDGC - Germany   \n",
              "..          ...                     ...                             ...   \n",
              "78   rs72712862      chr1:161000299:C:T  [EUR] UKBiobank - Case/Control   \n",
              "66   rs72712862      chr1:161000299:C:T           [EUR] IPDGC - UK GWAS   \n",
              "72   rs72712862      chr1:161000299:C:T                      [EUR] FIGS   \n",
              "108  rs72712862      chr1:161000299:C:T                    [ICE] deCODE   \n",
              "60   rs72712862      chr1:161000299:C:T          [EUR] IPDGC - TÃ¼bingen   \n",
              "\n",
              "         BETA         P        SE        OR       L95       U95  \n",
              "109 -0.057629  0.062097  0.030890  0.944000  0.888541  1.002920  \n",
              "27   0.076878  0.482758  0.109532  1.079910  0.871268  1.338515  \n",
              "38  -0.078313  0.336272  0.081444  0.924675  0.788248  1.084715  \n",
              "91   0.023228  0.780758  0.083455  1.023500  0.869060  1.205386  \n",
              "22  -0.065666  0.536008  0.106107  0.936444  0.760609  1.152928  \n",
              "..        ...       ...       ...       ...       ...       ...  \n",
              "78   0.239174  0.016003  0.099290  1.270200  1.045573  1.543085  \n",
              "66   0.200022  0.308008  0.196214  1.221430  0.831473  1.794275  \n",
              "72   0.244900  0.049260  0.124546  1.277494  1.000789  1.630703  \n",
              "108  0.371564  0.000516  0.107008  1.450001  1.175658  1.788361  \n",
              "60  -0.016584  0.975027  0.529763  0.983553  0.348221  2.778054  \n",
              "\n",
              "[122 rows x 9 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "from scipy.stats import norm\n",
        "\n",
        "# Dictionary mapping SNP identifiers to rsIDs\n",
        "snp_to_rsID = {\n",
        "    \"chr3:161328149:A:T\": \"rs11929238\",\n",
        "    \"chr16:52939375:A:G\": \"rs1971534\",\n",
        "    \"chr3:58410136:A:G\": \"rs56384862\",\n",
        "    \"chr1:161000299:C:T\": \"rs72712862\",\n",
        "    \"chr3:151990484:CGAAA:C\": \"rs10569457\",\n",
        "    \"chr6:29015989:G:A\": \"rs41316625\"\n",
        "}\n",
        "\n",
        "# Define the file paths for each cohort\n",
        "files_info = {\n",
        "    \"[EUR] GP2\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/GP2/POP_CONTROLS/GP2_EUR_CC_FOR_PLINK_MAF_AGE_POP_CONTROLS.txt\",\n",
        "    \"[EUR] IPDGC - Courage UK\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_COURAGE_UK_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - Dutch\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_DUTCH_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - Germany\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_GERMANY_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - McGill\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_MCGILL_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - MF\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_MF_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - NIA\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_NIA_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - Oslo\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_OSLO_EUR_CC_FOR_PLINK_PD.txt\",\n",
        "    \"[EUR] IPDGC - Shulman\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_SHULMAN_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - Spain\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_SPAIN3_EUR_CC_NO_AGE_COVAR_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - TÃ¼bingen\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_TUBI_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] IPDGC - UK GWAS\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_UK_GWAS_EUR_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] FIGS\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/METAS/FIGS_META_EUR_MAF_FOR_PLINK.txt\",\n",
        "    \"[EUR] UKBiobank - Case/Control\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/UKB/UKB_ALL/UKB_ALL_CC_FOR_PLINK_MAF\",\n",
        "    \"[AJ] GP2\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/GP2/UPDATED/GP2_AJ_CC_FOR_PLINK_MAF_update.txt\",\n",
        "    \"[AJ] FIGS\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/FIGS/FIGS_AJ_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[FIN] FinnGen - R11\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/FINNGEN_R10_PD/finngen_R11_G6_PARKINSON_FOR_PLINK\",\n",
        "    \"[FIN] IPDGC - Finland\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/IPDGC/IPDGC_FINLAND_FIN_CC_NO_AGE_FOR_PLINK.txt\",\n",
        "    \"[ICE] deCODE\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/deCODE/deCODE_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] Million Veteran Program\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/MVP_CC_FOR_PLINK_MAF.txt\",\n",
        "    \"[EUR] UKBiobank - Proxy Cases\": \"/data/CARD_AA/projects/2024_GP2_EUROPEAN_META/UKB/UKB_PROXY/UKB_PROXY_ADJUSTED_FOR_PLINK.txt\"\n",
        "}\n",
        "\n",
        "results = pd.DataFrame()\n",
        "\n",
        "# Process each file\n",
        "for cohort, file_path in files_info.items():\n",
        "    # Load data from each file (assuming tab-delimited with header)\n",
        "    df = pd.read_csv(file_path, delimiter='\\t')\n",
        "    \n",
        "    # Filter for each SNP in snp_data and calculate statistics\n",
        "    for _, snp in snp_data.iterrows():\n",
        "        snp_of_interest = snp['SNP']\n",
        "        filtered_data = df[df['SNP'] == snp_of_interest].copy()\n",
        "\n",
        "        if not filtered_data.empty:\n",
        "            # Calculate and set using loc to avoid SettingWithCopyWarning\n",
        "            filtered_data.loc[:, 'SE'] = np.abs(filtered_data['BETA'] / norm.ppf(filtered_data['P'] / 2))\n",
        "            filtered_data.loc[:, 'OR'] = np.exp(filtered_data['BETA'])\n",
        "            filtered_data.loc[:, 'L95'] = np.exp(filtered_data['BETA'] - 1.96 * filtered_data['SE'])\n",
        "            filtered_data.loc[:, 'U95'] = np.exp(filtered_data['BETA'] + 1.96 * filtered_data['SE'])\n",
        "\n",
        "            # Append the necessary information to the results DataFrame\n",
        "            for _, row in filtered_data.iterrows():\n",
        "                rsID = snp_to_rsID.get(snp_of_interest, \"Unknown\")  # Default to \"Unknown\" if SNP not in dictionary\n",
        "                results = results.append({\n",
        "                    \"rsID\": rsID,\n",
        "                    \"SNP(hg38)\": snp_of_interest,\n",
        "                    \"Cohort\": cohort,\n",
        "                    \"BETA\": row['BETA'],\n",
        "                    \"P\": row['P'],\n",
        "                    \"SE\": row['SE'],\n",
        "                    \"OR\": row['OR'],\n",
        "                    \"L95\": row['L95'],\n",
        "                    \"U95\": row['U95']\n",
        "                }, ignore_index=True)\n",
        "\n",
        "\n",
        "results = results.sort_values(by='rsID')\n",
        "display(results)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "results.to_csv(f'{WORK_DIR}/METAS/FINAL_SETS/FOREST_PLOTS/data/remaining_cojo_rsIDs_ORs.txt', index=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Make one big file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "cojo_ors_file1 = pd.read_csv(f'{WORK_DIR}/METAS/FINAL_SETS/FOREST_PLOTS/data/rsIDs_cojo_with_ORs_allCohorts.txt', sep=\"\\t\")\n",
        "cojo_ors_file2 = pd.read_csv(f'{WORK_DIR}/METAS/FINAL_SETS/FOREST_PLOTS/data/remaining_cojo_rsIDs_ORs.txt')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>rsID</th>\n",
              "      <th>SNP</th>\n",
              "      <th>Cohort</th>\n",
              "      <th>BETA</th>\n",
              "      <th>SE</th>\n",
              "      <th>P</th>\n",
              "      <th>OR</th>\n",
              "      <th>L95</th>\n",
              "      <th>U95</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[AJ] FIGS</td>\n",
              "      <td>0.050931</td>\n",
              "      <td>0.071870</td>\n",
              "      <td>0.478523</td>\n",
              "      <td>1.052250</td>\n",
              "      <td>0.913992</td>\n",
              "      <td>1.211422</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[EUR] FIGS</td>\n",
              "      <td>-0.051600</td>\n",
              "      <td>0.020222</td>\n",
              "      <td>0.010720</td>\n",
              "      <td>0.949709</td>\n",
              "      <td>0.912803</td>\n",
              "      <td>0.988106</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[AJ] GP2</td>\n",
              "      <td>-0.101593</td>\n",
              "      <td>0.082202</td>\n",
              "      <td>0.216497</td>\n",
              "      <td>0.903397</td>\n",
              "      <td>0.768966</td>\n",
              "      <td>1.061330</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[EUR] GP2</td>\n",
              "      <td>-0.060258</td>\n",
              "      <td>0.015734</td>\n",
              "      <td>0.000128</td>\n",
              "      <td>0.941522</td>\n",
              "      <td>0.912930</td>\n",
              "      <td>0.971009</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[EUR] IPDGC - Courage UK</td>\n",
              "      <td>-0.127412</td>\n",
              "      <td>0.075081</td>\n",
              "      <td>0.089698</td>\n",
              "      <td>0.880371</td>\n",
              "      <td>0.759899</td>\n",
              "      <td>1.019943</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3318</th>\n",
              "      <td>rs8130097</td>\n",
              "      <td>chr21:45161847:C:A</td>\n",
              "      <td>[EUR] IPDGC - UK GWAS</td>\n",
              "      <td>0.081101</td>\n",
              "      <td>0.093659</td>\n",
              "      <td>0.386531</td>\n",
              "      <td>1.084480</td>\n",
              "      <td>0.902604</td>\n",
              "      <td>1.303005</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3320</th>\n",
              "      <td>rs8130097</td>\n",
              "      <td>chr21:45161847:C:A</td>\n",
              "      <td>[EUR] Million Veteran Program</td>\n",
              "      <td>0.058583</td>\n",
              "      <td>0.039285</td>\n",
              "      <td>0.135900</td>\n",
              "      <td>1.060333</td>\n",
              "      <td>0.981753</td>\n",
              "      <td>1.145202</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3321</th>\n",
              "      <td>rs8130097</td>\n",
              "      <td>chr21:45161847:C:A</td>\n",
              "      <td>[EUR] UKBiobank - Case/Control</td>\n",
              "      <td>0.147687</td>\n",
              "      <td>0.051833</td>\n",
              "      <td>0.004381</td>\n",
              "      <td>1.159150</td>\n",
              "      <td>1.047173</td>\n",
              "      <td>1.283101</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3322</th>\n",
              "      <td>rs8130097</td>\n",
              "      <td>chr21:45161847:C:A</td>\n",
              "      <td>[EUR] UKBiobank - Proxy Cases</td>\n",
              "      <td>0.089411</td>\n",
              "      <td>0.051567</td>\n",
              "      <td>0.082937</td>\n",
              "      <td>1.093530</td>\n",
              "      <td>0.988408</td>\n",
              "      <td>1.209833</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3323</th>\n",
              "      <td>rs8130097</td>\n",
              "      <td>chr21:45161847:C:A</td>\n",
              "      <td>[FIN] FinnGen - R11</td>\n",
              "      <td>0.042397</td>\n",
              "      <td>0.039599</td>\n",
              "      <td>0.284322</td>\n",
              "      <td>1.043309</td>\n",
              "      <td>0.965396</td>\n",
              "      <td>1.127510</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3169 rows Ã— 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "           rsID                 SNP                          Cohort      BETA  \\\n",
              "0      rs377808   chr1:52724687:T:G                       [AJ] FIGS  0.050931   \n",
              "1      rs377808   chr1:52724687:T:G                      [EUR] FIGS -0.051600   \n",
              "2      rs377808   chr1:52724687:T:G                        [AJ] GP2 -0.101593   \n",
              "3      rs377808   chr1:52724687:T:G                       [EUR] GP2 -0.060258   \n",
              "4      rs377808   chr1:52724687:T:G        [EUR] IPDGC - Courage UK -0.127412   \n",
              "...         ...                 ...                             ...       ...   \n",
              "3318  rs8130097  chr21:45161847:C:A           [EUR] IPDGC - UK GWAS  0.081101   \n",
              "3320  rs8130097  chr21:45161847:C:A   [EUR] Million Veteran Program  0.058583   \n",
              "3321  rs8130097  chr21:45161847:C:A  [EUR] UKBiobank - Case/Control  0.147687   \n",
              "3322  rs8130097  chr21:45161847:C:A   [EUR] UKBiobank - Proxy Cases  0.089411   \n",
              "3323  rs8130097  chr21:45161847:C:A             [FIN] FinnGen - R11  0.042397   \n",
              "\n",
              "            SE         P        OR       L95       U95  \n",
              "0     0.071870  0.478523  1.052250  0.913992  1.211422  \n",
              "1     0.020222  0.010720  0.949709  0.912803  0.988106  \n",
              "2     0.082202  0.216497  0.903397  0.768966  1.061330  \n",
              "3     0.015734  0.000128  0.941522  0.912930  0.971009  \n",
              "4     0.075081  0.089698  0.880371  0.759899  1.019943  \n",
              "...        ...       ...       ...       ...       ...  \n",
              "3318  0.093659  0.386531  1.084480  0.902604  1.303005  \n",
              "3320  0.039285  0.135900  1.060333  0.981753  1.145202  \n",
              "3321  0.051833  0.004381  1.159150  1.047173  1.283101  \n",
              "3322  0.051567  0.082937  1.093530  0.988408  1.209833  \n",
              "3323  0.039599  0.284322  1.043309  0.965396  1.127510  \n",
              "\n",
              "[3169 rows x 9 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "cojo_ors_file1 = cojo_ors_file1[cojo_ors_file1['Cohort'] != '[TOTAL] Meta-analysis']\n",
        "cojo_ors_file1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>rsID</th>\n",
              "      <th>SNP</th>\n",
              "      <th>Cohort</th>\n",
              "      <th>BETA</th>\n",
              "      <th>SE</th>\n",
              "      <th>P</th>\n",
              "      <th>OR</th>\n",
              "      <th>L95</th>\n",
              "      <th>U95</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[ICE] deCODE</td>\n",
              "      <td>-0.057629</td>\n",
              "      <td>0.030890</td>\n",
              "      <td>0.062097</td>\n",
              "      <td>0.944000</td>\n",
              "      <td>0.888541</td>\n",
              "      <td>1.002920</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[EUR] IPDGC - McGill</td>\n",
              "      <td>0.076878</td>\n",
              "      <td>0.109532</td>\n",
              "      <td>0.482758</td>\n",
              "      <td>1.079910</td>\n",
              "      <td>0.871268</td>\n",
              "      <td>1.338515</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[EUR] IPDGC - NIA</td>\n",
              "      <td>-0.078313</td>\n",
              "      <td>0.081444</td>\n",
              "      <td>0.336272</td>\n",
              "      <td>0.924675</td>\n",
              "      <td>0.788248</td>\n",
              "      <td>1.084715</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[AJ] FIGS</td>\n",
              "      <td>0.023228</td>\n",
              "      <td>0.083455</td>\n",
              "      <td>0.780758</td>\n",
              "      <td>1.023500</td>\n",
              "      <td>0.869060</td>\n",
              "      <td>1.205386</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rs10569457</td>\n",
              "      <td>chr3:151990484:CGAAA:C</td>\n",
              "      <td>[EUR] IPDGC - Germany</td>\n",
              "      <td>-0.065666</td>\n",
              "      <td>0.106107</td>\n",
              "      <td>0.536008</td>\n",
              "      <td>0.936444</td>\n",
              "      <td>0.760609</td>\n",
              "      <td>1.152928</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>117</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] UKBiobank - Case/Control</td>\n",
              "      <td>0.239174</td>\n",
              "      <td>0.099290</td>\n",
              "      <td>0.016003</td>\n",
              "      <td>1.270200</td>\n",
              "      <td>1.045573</td>\n",
              "      <td>1.543085</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>118</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] IPDGC - UK GWAS</td>\n",
              "      <td>0.200022</td>\n",
              "      <td>0.196214</td>\n",
              "      <td>0.308008</td>\n",
              "      <td>1.221430</td>\n",
              "      <td>0.831473</td>\n",
              "      <td>1.794275</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>119</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] FIGS</td>\n",
              "      <td>0.244900</td>\n",
              "      <td>0.124546</td>\n",
              "      <td>0.049260</td>\n",
              "      <td>1.277494</td>\n",
              "      <td>1.000789</td>\n",
              "      <td>1.630703</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>120</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[ICE] deCODE</td>\n",
              "      <td>0.371564</td>\n",
              "      <td>0.107008</td>\n",
              "      <td>0.000516</td>\n",
              "      <td>1.450001</td>\n",
              "      <td>1.175658</td>\n",
              "      <td>1.788361</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] IPDGC - TÃ¼bingen</td>\n",
              "      <td>-0.016584</td>\n",
              "      <td>0.529763</td>\n",
              "      <td>0.975027</td>\n",
              "      <td>0.983553</td>\n",
              "      <td>0.348221</td>\n",
              "      <td>2.778054</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>122 rows Ã— 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "           rsID                     SNP                          Cohort  \\\n",
              "0    rs10569457  chr3:151990484:CGAAA:C                    [ICE] deCODE   \n",
              "1    rs10569457  chr3:151990484:CGAAA:C            [EUR] IPDGC - McGill   \n",
              "2    rs10569457  chr3:151990484:CGAAA:C               [EUR] IPDGC - NIA   \n",
              "3    rs10569457  chr3:151990484:CGAAA:C                       [AJ] FIGS   \n",
              "4    rs10569457  chr3:151990484:CGAAA:C           [EUR] IPDGC - Germany   \n",
              "..          ...                     ...                             ...   \n",
              "117  rs72712862      chr1:161000299:C:T  [EUR] UKBiobank - Case/Control   \n",
              "118  rs72712862      chr1:161000299:C:T           [EUR] IPDGC - UK GWAS   \n",
              "119  rs72712862      chr1:161000299:C:T                      [EUR] FIGS   \n",
              "120  rs72712862      chr1:161000299:C:T                    [ICE] deCODE   \n",
              "121  rs72712862      chr1:161000299:C:T          [EUR] IPDGC - TÃ¼bingen   \n",
              "\n",
              "         BETA        SE         P        OR       L95       U95  \n",
              "0   -0.057629  0.030890  0.062097  0.944000  0.888541  1.002920  \n",
              "1    0.076878  0.109532  0.482758  1.079910  0.871268  1.338515  \n",
              "2   -0.078313  0.081444  0.336272  0.924675  0.788248  1.084715  \n",
              "3    0.023228  0.083455  0.780758  1.023500  0.869060  1.205386  \n",
              "4   -0.065666  0.106107  0.536008  0.936444  0.760609  1.152928  \n",
              "..        ...       ...       ...       ...       ...       ...  \n",
              "117  0.239174  0.099290  0.016003  1.270200  1.045573  1.543085  \n",
              "118  0.200022  0.196214  0.308008  1.221430  0.831473  1.794275  \n",
              "119  0.244900  0.124546  0.049260  1.277494  1.000789  1.630703  \n",
              "120  0.371564  0.107008  0.000516  1.450001  1.175658  1.788361  \n",
              "121 -0.016584  0.529763  0.975027  0.983553  0.348221  2.778054  \n",
              "\n",
              "[122 rows x 9 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "cojo_ors_file2\n",
        "cojo_ors_file2.rename(columns={\"SNP(hg38)\": \"SNP\"}, inplace=True)\n",
        "cojo_ors_file2_rerrange = cojo_ors_file2[['rsID', 'SNP', 'Cohort', 'BETA', 'SE', 'P', 'OR', 'L95', 'U95']].copy()\n",
        "cojo_ors_file2_rerrange"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>rsID</th>\n",
              "      <th>SNP</th>\n",
              "      <th>Cohort</th>\n",
              "      <th>BETA</th>\n",
              "      <th>SE</th>\n",
              "      <th>P</th>\n",
              "      <th>OR</th>\n",
              "      <th>L95</th>\n",
              "      <th>U95</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[AJ] FIGS</td>\n",
              "      <td>0.050931</td>\n",
              "      <td>0.071870</td>\n",
              "      <td>0.478523</td>\n",
              "      <td>1.052250</td>\n",
              "      <td>0.913992</td>\n",
              "      <td>1.211422</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[EUR] FIGS</td>\n",
              "      <td>-0.051600</td>\n",
              "      <td>0.020222</td>\n",
              "      <td>0.010720</td>\n",
              "      <td>0.949709</td>\n",
              "      <td>0.912803</td>\n",
              "      <td>0.988106</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[AJ] GP2</td>\n",
              "      <td>-0.101593</td>\n",
              "      <td>0.082202</td>\n",
              "      <td>0.216497</td>\n",
              "      <td>0.903397</td>\n",
              "      <td>0.768966</td>\n",
              "      <td>1.061330</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[EUR] GP2</td>\n",
              "      <td>-0.060258</td>\n",
              "      <td>0.015734</td>\n",
              "      <td>0.000128</td>\n",
              "      <td>0.941522</td>\n",
              "      <td>0.912930</td>\n",
              "      <td>0.971009</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rs377808</td>\n",
              "      <td>chr1:52724687:T:G</td>\n",
              "      <td>[EUR] IPDGC - Courage UK</td>\n",
              "      <td>-0.127412</td>\n",
              "      <td>0.075081</td>\n",
              "      <td>0.089698</td>\n",
              "      <td>0.880371</td>\n",
              "      <td>0.759899</td>\n",
              "      <td>1.019943</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3285</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] UKBiobank - Case/Control</td>\n",
              "      <td>0.239174</td>\n",
              "      <td>0.099290</td>\n",
              "      <td>0.016003</td>\n",
              "      <td>1.270200</td>\n",
              "      <td>1.045573</td>\n",
              "      <td>1.543085</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3286</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] IPDGC - UK GWAS</td>\n",
              "      <td>0.200022</td>\n",
              "      <td>0.196214</td>\n",
              "      <td>0.308008</td>\n",
              "      <td>1.221430</td>\n",
              "      <td>0.831473</td>\n",
              "      <td>1.794275</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3287</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] FIGS</td>\n",
              "      <td>0.244900</td>\n",
              "      <td>0.124546</td>\n",
              "      <td>0.049260</td>\n",
              "      <td>1.277494</td>\n",
              "      <td>1.000789</td>\n",
              "      <td>1.630703</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3288</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[ICE] deCODE</td>\n",
              "      <td>0.371564</td>\n",
              "      <td>0.107008</td>\n",
              "      <td>0.000516</td>\n",
              "      <td>1.450001</td>\n",
              "      <td>1.175658</td>\n",
              "      <td>1.788361</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3289</th>\n",
              "      <td>rs72712862</td>\n",
              "      <td>chr1:161000299:C:T</td>\n",
              "      <td>[EUR] IPDGC - TÃ¼bingen</td>\n",
              "      <td>-0.016584</td>\n",
              "      <td>0.529763</td>\n",
              "      <td>0.975027</td>\n",
              "      <td>0.983553</td>\n",
              "      <td>0.348221</td>\n",
              "      <td>2.778054</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3290 rows Ã— 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "            rsID                 SNP                          Cohort  \\\n",
              "0       rs377808   chr1:52724687:T:G                       [AJ] FIGS   \n",
              "1       rs377808   chr1:52724687:T:G                      [EUR] FIGS   \n",
              "2       rs377808   chr1:52724687:T:G                        [AJ] GP2   \n",
              "3       rs377808   chr1:52724687:T:G                       [EUR] GP2   \n",
              "4       rs377808   chr1:52724687:T:G        [EUR] IPDGC - Courage UK   \n",
              "...          ...                 ...                             ...   \n",
              "3285  rs72712862  chr1:161000299:C:T  [EUR] UKBiobank - Case/Control   \n",
              "3286  rs72712862  chr1:161000299:C:T           [EUR] IPDGC - UK GWAS   \n",
              "3287  rs72712862  chr1:161000299:C:T                      [EUR] FIGS   \n",
              "3288  rs72712862  chr1:161000299:C:T                    [ICE] deCODE   \n",
              "3289  rs72712862  chr1:161000299:C:T          [EUR] IPDGC - TÃ¼bingen   \n",
              "\n",
              "          BETA        SE         P        OR       L95       U95  \n",
              "0     0.050931  0.071870  0.478523  1.052250  0.913992  1.211422  \n",
              "1    -0.051600  0.020222  0.010720  0.949709  0.912803  0.988106  \n",
              "2    -0.101593  0.082202  0.216497  0.903397  0.768966  1.061330  \n",
              "3    -0.060258  0.015734  0.000128  0.941522  0.912930  0.971009  \n",
              "4    -0.127412  0.075081  0.089698  0.880371  0.759899  1.019943  \n",
              "...        ...       ...       ...       ...       ...       ...  \n",
              "3285  0.239174  0.099290  0.016003  1.270200  1.045573  1.543085  \n",
              "3286  0.200022  0.196214  0.308008  1.221430  0.831473  1.794275  \n",
              "3287  0.244900  0.124546  0.049260  1.277494  1.000789  1.630703  \n",
              "3288  0.371564  0.107008  0.000516  1.450001  1.175658  1.788361  \n",
              "3289 -0.016584  0.529763  0.975027  0.983553  0.348221  2.778054  \n",
              "\n",
              "[3290 rows x 9 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "cojo_ors_file2_rerrange = cojo_ors_file2_rerrange.iloc[1:]\n",
        "\n",
        "merged_df = pd.concat([cojo_ors_file1, cojo_ors_file2_rerrange], ignore_index=True)\n",
        "merged_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "merged_df.to_csv(f'{WORK_DIR}/METAS/FINAL_SETS/FOREST_PLOTS/data/rsIDs_cojo_with_ORs_allCohorts_DEC2024.txt', index=False, sep=\"\\t\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SNP\tCHR\tBP\tA1\tA2\tBETA\tSE\tP\tNMISS\n",
            "chr1:161000299:C:T\t1\t161000299\tT\tC\t0.23917436599731445\t0.09928940236568451\t0.0160034\t15455.482915610626\n"
          ]
        }
      ],
      "source": [
        "## spot check a few \n",
        "! head -1 {WORK_DIR}/UKB/UKB_ALL/UKB_ALL_CC_FOR_PLINK_MAF\n",
        "! grep \"chr1:161000299:C:T\" {WORK_DIR}/UKB/UKB_ALL/UKB_ALL_CC_FOR_PLINK_MAF"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SNP\tCHR\tBP\tA1\tA2\tBETA\tSE\tP\tNMISS\n",
            "chr21:45161847:C:A\t21\t45161847\tA\tC\t0.058582957834005356\t0.039284877479076385\t0.1359\t34141.015226436444\n"
          ]
        }
      ],
      "source": [
        "## spot check a few \n",
        "! head -1 {WORK_DIR}/MVP_CC_FOR_PLINK_MAF.txt\n",
        "! grep \"chr21:45161847:C:A\" {WORK_DIR}/MVP_CC_FOR_PLINK_MAF.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SNP\tCHR\tBP\tA1\tA2\tBETA\tSE\tP\tNMISS\n",
            "chr12:40340400:G:A\t12\t40340400\tA\tG\t1.6378374099731445\t0.24285557866096497\t1.54e-11\t34141.015226436444\n"
          ]
        }
      ],
      "source": [
        "## spot check a few \n",
        "! head -1 {WORK_DIR}/MVP_CC_FOR_PLINK_MAF.txt\n",
        "! grep \"chr12:40340400\" {WORK_DIR}/MVP_CC_FOR_PLINK_MAF.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Files have been successfully created.\n"
          ]
        }
      ],
      "source": [
        "grouped = merged_df.groupby('rsID')\n",
        "\n",
        "for rsID, group in grouped:\n",
        "    filename = f\"{WORK_DIR}/METAS/FINAL_SETS/FOREST_PLOTS/data/rsID_files_withMeta_cojo/{rsID}_split.txt\"\n",
        "    group.to_csv(filename, sep='\\t', index=False)\n",
        "\n",
        "print(\"Files have been successfully created.\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Example - Forest Plot code\n",
        "in R"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Required Libraries\n",
        "library(metafor)\n",
        "\n",
        "# Function to clean filenames by removing problematic characters\n",
        "clean_filename <- function(filename) {\n",
        "  # Remove leading/trailing whitespace\n",
        "  filename <- trimws(filename)\n",
        "  # Replace spaces with underscores\n",
        "  filename <- gsub(\" \", \"_\", filename)\n",
        "  # Remove single quotes\n",
        "  filename <- gsub(\"'\", \"\", filename)\n",
        "  # Return cleaned filename\n",
        "  return(filename)\n",
        "}\n",
        "\n",
        "# Read in rsID and nearest gene information\n",
        "rsid_gene_info <- read.table(\"./METAS/FINAL_SETS/FOREST_PLOTS/data/rsID_and_nearest_gene.txt\", header = TRUE, sep = \"\\t\")\n",
        "\n",
        "# Directories\n",
        "input_directory <- \"./METAS/FINAL_SETS/FOREST_PLOTS/data/rsID_files_withMeta_cojo/\"\n",
        "output_directory <- \"./METAS/FINAL_SETS/FOREST_PLOTS/data/cojo_forest_plots/\"\n",
        "error_log_file <- \"./METAS/FINAL_SETS/FOREST_PLOTS/data/error_log.txt\"\n",
        "\n",
        "dir.create(output_directory, showWarnings = FALSE)\n",
        "\n",
        "# Initialize error log\n",
        "write.table(data.frame(rsID = character(), Cohort = character(), Error = character()),\n",
        "            file = error_log_file, sep = \"\\t\", row.names = FALSE, col.names = TRUE, quote = FALSE)\n",
        "\n",
        "# Files\n",
        "files <- list.files(input_directory, pattern = \".txt$\", full.names = TRUE)\n",
        "cohort_order_path <- \"./METAS/FINAL_SETS/FOREST_PLOTS/data/cohort_names.txt\"\n",
        "cohort_order <- read.table(cohort_order_path, header = TRUE, sep = \"\\t\")\n",
        "\n",
        "# Loop\n",
        "for (file_path in files) {\n",
        "  print(paste(\"Processing file:\", file_path))\n",
        "  \n",
        "  # Read in data\n",
        "  data <- read.table(file_path, header = TRUE, sep = \"\\t\")\n",
        "  \n",
        "  # Check for empty or invalid data\n",
        "  if (nrow(data) == 0 || all(is.na(data))) {\n",
        "    print(paste(\"Skipping file due to invalid data:\", file_path))\n",
        "    next\n",
        "  }\n",
        "  \n",
        "  # Sort by cohort order\n",
        "  data <- data[match(cohort_order$Rename, data$Cohort), ]\n",
        "  data <- na.omit(data)\n",
        "  \n",
        "  # Drop cohorts where BETA is zero and log the cohort in the error log\n",
        "  zero_beta_cohorts <- data$Cohort[data$BETA == 0]\n",
        "  if (length(zero_beta_cohorts) > 0) {\n",
        "    for (cohort in zero_beta_cohorts) {\n",
        "      rsID <- ifelse(nrow(data) > 0, as.character(data$rsID[1]), \"Unknown\")\n",
        "      write.table(data.frame(rsID = rsID, Cohort = cohort, Error = \"BETA is zero\"),\n",
        "                  file = error_log_file, sep = \"\\t\", row.names = FALSE, col.names = FALSE, append = TRUE, quote = FALSE)\n",
        "    }\n",
        "    data <- data[data$BETA != 0, ]\n",
        "  }\n",
        "  \n",
        "  # Check for missing values in key columns\n",
        "  if (any(is.na(data$BETA) | is.na(data$SE))) {\n",
        "    print(paste(\"Skipping file due to NA values in BETA or SE:\", file_path))\n",
        "    next\n",
        "  }\n",
        "  \n",
        "  # Analysis\n",
        "  labs_normal <- data$Cohort\n",
        "  resFe_normal <- tryCatch({\n",
        "    rma(yi = data$BETA, sei = data$SE, slab = labs_normal, method = \"FE\")\n",
        "  }, error = function(e) {\n",
        "    # Log the error and rsID to the error log file\n",
        "    rsID <- ifelse(nrow(data) > 0, as.character(data$rsID[1]), \"Unknown\")\n",
        "    write.table(data.frame(rsID = rsID, Cohort = \"All\", Error = e$message),\n",
        "                file = error_log_file, sep = \"\\t\", row.names = FALSE, col.names = FALSE, append = TRUE, quote = FALSE)\n",
        "    print(paste(\"Error in rma for file:\", file_path, \" - \", e$message))\n",
        "    return(NULL) # Skip to next iteration\n",
        "  })\n",
        "  \n",
        "  # Skip plotting if `rma` failed\n",
        "  if (is.null(resFe_normal)) {\n",
        "    next\n",
        "  }\n",
        "  \n",
        "  # Extract rsID and nearest gene for plot title\n",
        "  rsid <- as.character(data$rsID[1])\n",
        "  nearest_gene <- rsid_gene_info$Nearest_Gene[match(rsid, rsid_gene_info$rsID)]\n",
        "  plot_title <- paste(rsid, \"- Nearest Gene:\", nearest_gene)\n",
        "  \n",
        "  # Plot\n",
        "  output_file <- paste(output_directory, clean_filename(rsid), \".png\", sep = \"\")\n",
        "  png(file = output_file, width = 800, height = 700)\n",
        "  \n",
        "  Pvalue_normal <- formatC(resFe_normal$pval, digits = 4)\n",
        "  options(na.action = \"na.pass\")\n",
        "  par(mar = c(5, 4, 1, 1))\n",
        "  \n",
        "  forest(resFe_normal, annotate = TRUE, xlim = c(-2.25, 3.25), width = 3, cex.lab = 0.8, cex.axis = 1,\n",
        "         atransf = exp, xlab = \"Odds Ratio (95%CI)\", slab = labs_normal, mlab = \"[TOTAL] Meta-analysis\",\n",
        "         col = \"red\", border = \"red\", cex = 0.9, at = log(c(0.5, 1, 2, 3)))\n",
        "  text(0, length(labs_normal) + 2, plot_title, cex = 1.5, font = 2)\n",
        "  \n",
        "  dev.off()\n",
        "}"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "machine_shape": "hm",
      "provenance": [],
      "toc_visible": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
